src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Query/LaunchProfiles/DuplicateLaunchProfileAction.cs

```diff
@@ -0,0 +1,41 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+using Microsoft.VisualStudio.ProjectSystem.Query;
+using Microsoft.VisualStudio.ProjectSystem.Query.ProjectModelMethods.Actions;
+using Microsoft.VisualStudio.ProjectSystem.Query.QueryExecution;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Query
+{
+    /// <summary>
+    /// <see cref="IQueryActionExecutor"/> handling <see cref="ProjectModelActionNames.DuplicateLaunchProfile"/> actions.
+    /// </summary>
+    internal sealed class DuplicateLaunchProfileAction : LaunchProfileActionBase
+    {
+        private readonly DuplicateLaunchProfile _executableStep;
+
+        public DuplicateLaunchProfileAction(DuplicateLaunchProfile executableStep)
+        {
+            _executableStep = executableStep;
+        }
+
+        protected override async Task ExecuteAsync(ILaunchSettingsProvider launchSettingsProvider)
+        {
+            ILaunchSettings? launchSettings = await launchSettingsProvider.WaitForFirstSnapshot(Timeout.Infinite);
```

Comment:
...and use `CancellationToken` here, for example.

Reply:
You mean pass the token to `WaitForFirstSnapshot` or check it for cancellation?

Reply:
Pass it, if accepted, or use `.WithCancellation` or whatever the extension method is called.

Reply:
`WaitForFirstSnapshot` takes a time out (in milliseconds) rather than a `CancellationToken`.

Reply:
It's the infinite timeout that concerns me here. If that snapshot never arrives (for one of many reasons) any tasks waiting on this to complete will linger indefinitely.

Looking at the implementation in `LaunchSettingsProvider`, I think it should call `TryCancel` on its `_firstSnapshotCompletionSource` in its `Dispose` method. Then if the provider is disposed while something is waiting for that first snapshot, the task will be cancelled and a leak will be avoided.

Reply:
https://github.com/dotnet/project-system/pull/7055 addresses part of my concern here.

The other part: if the execution is cancelled, the caller may still be stuck here waiting for longer than necessary until their cancellation takes effect. If it's not possible to modify `ILaunchSettingsProvider` then consider adding `ILaunchSettingsProvider2` that accepts a cancellation token. The current implementation can trivially be adapted to support cancellation.

Reply:
I think there's a bigger issue here in that `ILaunchSettingsProvider` in general has no support for cancellation. I've filed #7056 to address that.